src/component_selector.jl

```diff
@@ -0,0 +1,295 @@
+# ABSTRACT TYPE DEFINITIONS
+#=
+`ComponentSelector` extension notes:
+Concrete subtypes MUST implement:
+ - `get_components`: returns an iterable of components
+ - `get_name`: returns a name for the selector -- or use the default by implementing
+   `default_name` and having a `name` field
+ - `get_groups`: returns an iterable of `ComponentSelector`s
+
+Concrete subtypes SHOULD implement:
+ - The factory method `make_selector` (make sure your signature does not conflict with an
+   existing one)
+
+Concrete subtypes MAY implement:
+ - `default_name`
+=#
+"The base type for all `ComponentSelector`s."
+abstract type ComponentSelector end
+
+#=
+`SingularComponentSelector` extension notes:
+The interface is the same as for `ComponentSelector` except:
+ - `get_components` MUST return zero or one components
+ - the additional method `get_component` is part of the interface, but the default
+   implementation just wraps `get_components` and SHOULD NOT need to be overridden
+ - there is a sensible default for `get_groups`; it MAY be overridden
+=#
+"`ComponentSelector`s that can only refer to zero or one components."
+abstract type SingularComponentSelector <: ComponentSelector end
+
+#=
+`PluralComponentSelector` extension notes:
+The interface is the same as for `ComponentSelector`.
+=#
+"""`ComponentSelector`s that may refer to multiple components."""
+abstract type PluralComponentSelector <: ComponentSelector end
+
+#=
+`DynamicallyGroupedComponentSelector` extension notes:
+One MAY subtype this and have a `groupby::Union{Symbol, Function}` field to get an automatic
+implementation of `get_groups`.
+=#
+"`PluralComponentSelector`s whose grouping is determined by a `groupby` field."
+abstract type DynamicallyGroupedComponentSelector <: PluralComponentSelector end
```

Comment:
Dynamic roughly means that grouping is on the fly: when you call `get_groups`, it will take stock of what components actually exist in the system and make the groups as necessary. `ListComponentSelector` has groups, but they're not dynamic: if you construct a `ListComponentSelector` with two groups, it will always have those two groups no matter the system.

(in fact, it would be nice to be able to call `get_groups` on non-dynamically-grouped selectors *without* having to pass in the system, but that's for the next version to worry about)